python/lsst/drp/tasks/wcsfit.py

```diff
@@ -0,0 +1,1019 @@
+
+        Returns
+        -------
+        result : `lsst.pipe.base.Struct`
+            ``outputWCSs`` : `list` of `lsst.afw.table.ExposureCatalog`
+                List of exposure catalogs (one per visit) with the WCS for each
+                detector set by the new fitted WCS
+            ``fitModel`` : `wcsfit.WCSFit`
+                Model-fitting object with final model parameters
+            ``outputCatalog`` : `pd.DataFrame`
+                Catalog with fit residuals of all sources used
+        """
+        self.log.info("Gathering instrument, exposure, and field info")
+        # Set up an instrument object
+        instrument = wcsfit.Instrument(instrumentName)
+
+        # Get RA, Dec, MJD, etc., for the input visits
+        exposuresHelper, medianEpoch = self._get_exposure_info(inputVisitSummary, instrument,
+                                                               refEpoch=refEpoch)
+
+        # Get information about the extent of the input visits
+        fields, fieldCenter, fieldRadius = self._prep_sky(inputVisitSummary, medianEpoch)
+
+        self.log.info("Load catalogs and associate sources")
+        # Set up class to associate sources into matches using a
+        # friends-of-friends algorithm
+        associations = wcsfit.FoFClass(fields, [instrument], exposuresHelper,
+                                       [fieldRadius.asDegrees()],
+                                       (self.config.matchRadius * u.arcsec).to(u.degree).value)
+
+        # Add the reference catalog to the associator
+        self._load_refCat(associations, fieldCenter, fieldRadius, epoch=refEpoch)
+
+        # Add the science catalogs and associate new sources as they are added
+        self._load_catalogs_and_associate(associations, inputCatalogRefs)
+
+        self.log.info("Fit the WCSs")
+        # Set up a YAML-type string using the config variables and a sample
+        # visit
+        inputYAML = self.makeYAML(inputVisitSummary[0])
+
+        # Set up the WCS-fitting class using the results of the FOF associator
+        wcsf = wcsfit.WCSFit(fields, [instrument], exposuresHelper,
+                             self.extensionInfo['visitIndex'], self.extensionInfo['detectorIndex'],
+                             inputYAML, self.extensionInfo['wcs'], associations.sequence, associations.extn,
+                             associations.obj, sysErr=self.config.systematicError,
+                             refSysErr=self.config.referenceSystematicError,
+                             usePM=self.config.useProperMotion,
+                             verbose=self.config.verbose)
+
+        # Add the science and reference sources
+        self._addObjects(wcsf, inputCatalogRefs)
+        self._addRefObjects(wcsf)
+
+        # Do the WCS fit
+        wcsf.fit()
+        self.log.info("WCS fitting done")
+
+        outputWCSs = self._makeOutputs(wcsf, inputVisitSummary[0])
+        outputCatalog = pd.DataFrame(wcsf.getOutputCatalog())
```

Comment:
This mimics the input catalogs ("preSourceTable_visit", which are also DataFrames. This output exists to facilitate doing diagnostics, and is otherwise not used downstream.

Reply:
How about using `lsst.daf.butler.formatters.numpy_dict_to_arrow` since the original format is a dict of numpy arrays? And would the storage class be "ArrowNumpy" in that case?

Reply:
If you manually convert from numpy dict to arrow, the storage class would be `ArrowTable` in that case since you have converted it to an arrow table already and are storing it as an arrow table.  (There actually isn't a `NumpyDictArrow` or equivalent storage class which would allow you to store the numpy dict directly).  Anyway, doing the `numpy_dict_to_arrow` conversion and storing as an `ArrowTable` is a good idea and would probably be the most direct way of doing this.

Reply:
It's `lsst.daf.butler.formatters.parquet.numpy_dict_to_arrow`. If we are using these functions in user code and not internally in storage class converters then I'm a bit worried about them being in the wrong place in daf_butler.

Reply:
@timj This is a bit off-topic, the numpy dict conversion code is used internally, it just isn't exposed as a first-class storage class.

Reply:
Yes, I know we are using them internally, my comment was that the helper functions aren't intended to be public API for task code.

Reply:
I have filed https://jira.lsstcorp.org/browse/DM-37279 and will add a direct `ArrowNumpyDict` storage class which should be used here when that's ready (soon!).

Reply:
A simpler alternative that doesn't require that ticket: if you don't have any string columns, then you can easily use `pyarrow.Table.from_dict()` to convert to an arrow table, and then use `ArrowTable` storage class.

Reply:
Right, I think we talked about doing this a while ago. This is the approach I'm going to take.